miner/tx_pool_w_accounts.go

```diff
@@ -0,0 +1,111 @@
+package miner
+
+import (
+	"bytes"
+	"encoding/binary"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/pending_txs"
+	"github.com/spacemeshos/sha256-simd"
+	"sort"
+	"sync"
+)
+
+type innerPool interface {
+	Get(id types.TransactionId) (types.AddressableSignedTransaction, error)
+	GetAllItems() []types.AddressableSignedTransaction
+	Put(id types.TransactionId, item *types.AddressableSignedTransaction)
+	Invalidate(id types.TransactionId)
+}
+
+type TxPoolWithAccounts struct {
+	innerPool
+	accounts      map[types.Address]*pending_txs.AccountPendingTxs
+	accountsMutex sync.RWMutex
+}
+
+func NewTxPoolWithAccounts() *TxPoolWithAccounts {
+	return &TxPoolWithAccounts{
+		innerPool: NewTxMemPool(),
+		accounts:  make(map[types.Address]*pending_txs.AccountPendingTxs),
+	}
+}
+
+func (t *TxPoolWithAccounts) GetRandomTxs(numOfTxs int, seed []byte) []types.AddressableSignedTransaction {
+	txs := t.innerPool.GetAllItems()
+	if len(txs) <= numOfTxs {
+		return txs
+	}
+
+	sort.Slice(txs, func(i, j int) bool {
+		id1 := types.GetTransactionId(txs[i].SerializableSignedTransaction)
+		id2 := types.GetTransactionId(txs[j].SerializableSignedTransaction)
+		return bytes.Compare(id1[:], id2[:]) < 0
+	})
+	var ret []types.AddressableSignedTransaction
+	for idx := range getRandIdxs(numOfTxs, len(txs), seed) {
+		ret = append(ret, txs[idx])
+	}
+	return ret
+}
+
+func getRandIdxs(numOfTxs, spaceSize int, seed []byte) map[uint64]struct{} {
+	idxs := make(map[uint64]struct{})
+	i := uint64(0)
+	for len(idxs) < numOfTxs {
+		message := make([]byte, len(seed)+binary.Size(i))
```

Comment:
maybe make the slice outside the loop save on mallocs

Reply:
done